### PR TITLE
Refactor action policy change handler

### PR DESCRIPTION
**Pull request by pchila**
## What does this PR do?
Refactor action policy handler splitting validation, persistence and runtime update of the new configuration

## Why is it important?

This is a prerequisite for implementing #2851

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works~~
- ~~[ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)~~
- ~~[ ] I have added an integration test or an E2E test~~

## Author's Checklist

- [ ]

## Related issues

- 

## Questions to ask yourself

- How are we going to support this in production? 
- How are we going to measure its adoption? 
- How are we going to debug this?
- What are the metrics I should take care of?
- ...

### CODE REVIEW

File: internal/pkg/agent/application/actions/handlers/handler_action_policy_change.go [hunk 1]
```diff
@@ -118,104 +118,160 @@ func (h *PolicyChangeHandler) Watch() <-chan coordinator.ConfigChange {
...
-	prevProtocol := h.config.Fleet.Client.Protocol
-	prevPath := h.config.Fleet.Client.Path
-	prevHost := h.config.Fleet.Client.Host
-	prevHosts := h.config.Fleet.Client.Hosts
-	prevProxy := h.config.Fleet.Client.Transport.Proxy
-	h.config.Fleet.Client.Protocol = cfg.Fleet.Client.Protocol
-	h.config.Fleet.Client.Path = cfg.Fleet.Client.Path
-	h.config.Fleet.Client.Host = cfg.Fleet.Client.Host
-	h.config.Fleet.Client.Hosts = cfg.Fleet.Client.Hosts
+	// make a copy the current client config and apply the changes in place on this copy
```
Comment (AndersonQ) [inline internal/pkg/agent/application/actions/handlers/handler_action_policy_change.go:132]: [Suggestion[
```suggestion
	// make a copy of the current client config and apply the changes in place to this this copy.
```

File: internal/pkg/agent/application/actions/handlers/handler_action_policy_change.go [hunk 2]
```diff
@@ -118,104 +118,160 @@ func (h *PolicyChangeHandler) Watch() <-chan coordinator.ConfigChange {
...
+	}
+
+	h.config.Fleet.Client = *validatedConfig
+
+	// the config has already been validated, no need for error handling
+	fleetClient, err := client.NewAuthWithConfig(
+		h.log, h.config.Fleet.AccessAPIKey, *validatedConfig)
+	if err != nil {
+		return fmt.Errorf("creating new fleet client with updated config: %w", err)
+	}
```
Comment (AndersonQ) [inline internal/pkg/agent/application/actions/handlers/handler_action_policy_change.go:250]: [Suggestion]
The error is being handles, so the comment should reflect it.

```suggestion
	// the config has already been validated, it should not error.
	fleetClient, err := client.NewAuthWithConfig(
		h.log, h.config.Fleet.AccessAPIKey, *validatedConfig)
	if err != nil {
		return fmt.Errorf("creating new fleet client with updated config: %w", err)
	}
```

**Review by AndersonQ — APPROVED** [review]

File: internal/pkg/agent/application/actions/handlers/handler_action_policy_change.go [hunk 3]
```diff
@@ -118,104 +118,160 @@ func (h *PolicyChangeHandler) Watch() <-chan coordinator.ConfigChange {
...
-	defer func() {
-		if err != nil {
-			h.config.Fleet.Client.Protocol = prevProtocol
-			h.config.Fleet.Client.Path = prevPath
-			h.config.Fleet.Client.Host = prevHost
-			h.config.Fleet.Client.Hosts = prevHosts
-			h.config.Fleet.Client.Transport.Proxy = prevProxy
-		}
-	}()
+	// Test new config
```
Comment (AndersonQ) [inline internal/pkg/agent/application/actions/handlers/handler_action_policy_change.go:136]: [Suggestion]
```suggestion
	// Test the new config
```

**Comment by pchila** [comment]
Fixed bug that caused Fleet APIKey to be removed from fleet.enc store https://buildkite.com/elastic/elastic-agent/builds/8306#018ed0f1-7e0d-4bbc-b906-28be625f219e

Fix in 237bc4a2fd772fc94638c2827a9f82fb7219c40f

**Comment by cmacknz** [comment]
> Fixed bug that caused Fleet APIKey to be removed from fleet.enc store https://buildkite.com/elastic/elastic-agent/builds/8306#018ed0f1-7e0d-4bbc-b906-28be625f219e

Is there a unit test that could have caught this?

**Comment by pchila** [comment]
> > Fixed bug that caused Fleet APIKey to be removed from fleet.enc store https://buildkite.com/elastic/elastic-agent/builds/8306#018ed0f1-7e0d-4bbc-b906-28be625f219e
> 
> Is there a unit test that could have caught this?

Not one that we have right now because the changes in memory actually worked correctly. The only way to detect this would be to write the config in the store and then either assert the contents of the store (probably the preferred option) or make the agent reload the config from file and attempt to contact fleet server

**Pull request merged**